File: 292.nim-game.cpp
/*
 * @lc app=leetcode id=292 lang=cpp
 *
 * [292] Nim Game
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  bool canWinNim(int n) {
    return n % 4 != 0;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.canWinNim(1), true);
  ASSERT_EQ(s.canWinNim(2), true);
  ASSERT_EQ(s.canWinNim(3), true);
  ASSERT_EQ(s.canWinNim(4), false);
  ASSERT_EQ(s.canWinNim(5), true);
  ASSERT_EQ(s.canWinNim(6), true);
  ASSERT_EQ(s.canWinNim(7), true);
  ASSERT_EQ(s.canWinNim(8), false);
  return 0;
}

File: 292.nim-game.rs
/*
 * @lc app=leetcode id=292 lang=rust
 *
 * [292] Nim Game
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn can_win_nim(n: i32) -> bool {
        n % 4 != 0
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::can_win_nim(1));
    assert!(Solution::can_win_nim(2));
    assert!(Solution::can_win_nim(3));
    assert!(!Solution::can_win_nim(4));
    assert!(Solution::can_win_nim(5));
    assert!(Solution::can_win_nim(6));
    assert!(Solution::can_win_nim(7));
    assert!(!Solution::can_win_nim(8));
}
